File: samples/c_samples/bench_primes.c
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>

#include "common.h"

static struct profile_info count_primes_info;

uint64_t count_primes(uint64_t limit) {
  uint64_t clock_start = clock_nanos();

  uint64_t accum = 0;
  for (uint64_t n = limit; n > 1; n--) {
    bool composite = false;
    for (uint64_t i = 2; i * i <= n; i++) {
      if (n % i == 0) {
        composite = true;
        break;
      }
    }
    if (!composite) {
      accum++;
    }
  }

  uint64_t clock_end = clock_nanos();
  record_call(&count_primes_info, clock_start, clock_end);

  return accum;
}

int main(int argc, char **argv) {
  unsigned int iters;
  unsigned int limit;
  if (scanf("%u\n%u", &iters, &limit) == EOF) {
    fprintf(stderr, "Expected two positive integers (an iteration count and a limit)\n");
    return 1;
  }

  if (iters == 0) {
    return 0;
  }

  uint64_t result;
  for (unsigned int i = 0; i < iters; i++) {
    result = count_primes(limit);
  }

  printf("There are %llu primes <= %u\n", (unsigned long long)result, limit);

  write_report(count_primes_info);
}

File: samples/rust_samples/bench_primes.rs
mod common;

use common::{read_line, repeat, write_report, ProfileInfo};

static COUNT_PRIMES_INFO: ProfileInfo = ProfileInfo::new();

fn count_primes(limit: u64) -> u64 {
    COUNT_PRIMES_INFO.record_call(|| {
        let mut accum = 0;
        for n in (2..limit).rev() {
            let mut composite = false;
            let mut i = 2;
            while i * i <= n {
                if n % i == 0 {
                    composite = true;
                    break;
                }
                i += 1;
            }
            if !composite {
                accum += 1;
            }
        }
        accum
    })
}

fn main() {
    match (read_line().parse::<u64>(), read_line().parse::<u64>()) {
        (Ok(iters), Ok(limit)) => {
            if let Some(result) = repeat(iters, || count_primes(limit)) {
                println!("There are {} primes <= {}", result, limit);
            }
        }

        (_, _) => {
            eprintln!("Please enter two positive integers (an iteration count and a limit)");
        }
    }

    write_report(&COUNT_PRIMES_INFO);
}
